IntegrationTests/UI.IntegrationTests/UITest.cs

```diff
@@ -27,32 +28,69 @@ void HandleApplicationIdle(object sender, EventArgs e)
-            // tracked by its collection. Otherwise, test code would not be able to wait for pending operations to
-            // complete.
-            var test = ThreadHelper.JoinableTaskContext.Factory.RunAsync(async () =>
+            // If form.Dipose was called by the async test task, closing the window would be done in a strange order.
+            T form = null;
+            try
             {
-                await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
-                await WaitForIdleAsync();
-                var dialog = Application.OpenForms.OfType<T>().Single();
-                try
-                {
-                    await runAsync(dialog);
-                }
-                finally
+                // Start runTestAsync before calling showForm.
+                // The latter might block until the form is closed, especially if using Application.Run(form).
+
+                // Avoid using ThreadHelper.JoinableTaskFactory for the outermost operation because we don't want the task
+                // tracked by its collection. Otherwise, test code would not be able to wait for pending operations to
+                // complete.
+                var test = ThreadHelper.JoinableTaskContext.Factory.RunAsync(async () =>
                 {
-                    dialog.Close();
-                }
-            });
+                    try
+                    {
+                        await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
+                        await WaitForIdleAsync();
+                        form = Application.OpenForms.OfType<T>().Single();
+                        await runTestAsync(form);
+                    }
+                    finally
+                    {
+                        try
+                        {
+                            form?.Close();
+                        }
+                        finally
+                        {
+                            // Try to close all open Forms in order to let return the blocking Application.Run(form)
+                            // which might have been called by showForm() below.
+                            for (int trial = 0; trial < 3 && Application.OpenForms.Count > 0; ++trial)
+                            {
+                                Application.DoEvents();
+                                Application.OpenForms.OfType<Form>().ForEach(f =>
+                                {
+                                    try
+                                    {
+                                        f.Close();
+                                    }
+                                    catch (Exception)
+                                    {
+                                        // ignore
+                                    }
+                                });
+                            }
```

Comment:
I don't understand why we need this block. I would prefer to remove it.

Reply:
If there is a _specific_ type of dialog we need to account for, I would prefer to have a more specific cleanup in place to handle it.

Reply:
Yes, this happens often for FormCommit. But yes, it has no effect on the test behavior. Then the loop always has all three iterations.

Reply:
For the FormCommit case, what form is open that needs to be closed?

Reply:
It is FormCommit itself that does not close.

Reply:
Though it has closed when `showForm` returns, i.e. weird but not causing the test to fail.

Reply:
removed

Reply:
Now I remember: I added it in order to avoid the hang caused by an exception before `form = OpenForms...`. Then I noticed that FormCommit trapped here, and I added the `for (int trial...)` loop, which did not change the behavior.
I have moved this up to the new `try catch` for the early exceptions.